File: 04_Stack/l001.java
class MyStack {

    private int[] arr = null;

    private int size = 0;

    private int tos = -1;

    public MyStack(int size) {
        this.arr = new int[size];
    }

    public int size() {
        return this.size;
    }

    public boolean isEmpty() {
        return this.size == 0;
    }

    public int top() {
        return this.arr[tos];
    }

    public void push(int data) {
        if (size > 10000) {
            System.out.println("StackOverFlow");
            return;
        }
        this.arr[size] = data;
        size++;
        tos++;
    }

    public int pop() {
        int data = this.arr[tos];
        this.arr[tos] = -1;
        size--;
        tos--;
        return data;
    }

    @Override
    public String toString() {
        String str = "[";
        int i = 0;
        for (; i < this.size - 1; i++) {
            str += this.arr[i] + ", ";
        }
        str += this.arr[i] + "]";
        return str;
    }
}

public class l001 {
    public static void main(String[] args) {
        MyStack st = new MyStack(10);

        for (int i = 0; i < 5; i++) {
            st.push(i + 10);
        }

        System.out.println(st);
        System.out.println(st.size());
        st.pop();
        System.out.println(st + "\n" + st.size());

        while (st.size() > 0) {
            int val = st.top();
            st.pop();
            System.out.println(val);
        }
    }
}

File: 04_Stack/l001.cpp
#include <iostream>

using namespace std;

class MyStack {
   private:
    int n = 10000;
    int* arr = nullptr;
    int sz = 0;
    int tos = -1;

   public:
    MyStack() {
        this->arr = new int[this->n];
    }

    MyStack(int n) {
        this->n = n;
        this->arr = new int[this->n];
    }

    void push(int val) {
        if (this->sz == this->n) return;
        this->tos++;
        this->arr[tos] = val;
        this->sz++;
    }

    void pop() {
        if (this->sz == 0) return;
        this->tos--;
        this->sz--;
    }

    int top() {
        return this->sz == 0 ? -1 : this->arr[tos];
    }

    int size() {
        return this->sz;
    }

    bool isEmpty() {
        return this->sz == 0;
    }
};

void solve() {
    MyStack st;

    for (int i = 1; i <= 10; i++) {
        st.push(10 + i);
    }

    while (st.size() > 0) {
        cout << st.top() << " ";
        st.pop();
    }
    // cout << st.top();
}

int main(int argc, char** argv) {
    solve();
}